File: 62-unique-paths/62-unique-paths.js
/**
 * @param {number} m
 * @param {number} n
 * @return {number}
 */

// DYNAMIC PROGRAMMING

// from each cell, num paths is down paths + right paths
// use tabulation => end to start;
// base case is at end (1 paths) => cal rest off that
// start with bottom row, then next till top
// bottom row is all 1s, same with right column
const uniquePaths = function(m, n, memo = {}) {
    const key = `${m},${n}`;
    if (m === 1 && n === 1) return 1;
    if (`${n},${m}` in memo || key in memo) return memo[key] ?? memo[`${n},${m}`];
    if (!m || !n) return 0;
    
    memo[key] = uniquePaths(m-1,n,memo) + uniquePaths(m, n-1,memo);
    
    return memo[key];
};

File: medium/62-unique-paths/62-unique-paths.rb
# @param {Integer} m
# @param {Integer} n
# @return {Integer}
def unique_paths(m, n)
    dp = Array.new n, 1
    
    (m - 1).times do 
        row = [1]
        (1...n).each do |i|
            row << row[-1] + dp[i] 
        end
        
        dp = row
    end
    
    dp[-1]
end